File: src/main/kotlin/com/huawei/cangjie/psi/CjVisitor.java
package com.huawei.cangjie.psi;

import com.intellij.psi.PsiElementVisitor;
import org.jetbrains.annotations.NotNull;

public class CjVisitor<R, D> extends PsiElementVisitor {

    public R visitCjElement(@NotNull CjElement element, D data) {
        visitElement(element);
        return null;
    }

    public R visitSafeQualifiedExpression(@NotNull CjSafeQualifiedExpression expression, D data) {
        return visitQualifiedExpression(expression, data);
    }

    public R visitQuoteExpression(@NotNull CjQuoteExpression element, D data) {
        return visitExpression(element, data);


    }

    public R visitOptionType(@NotNull CjOptionType optionType, D data) {
        return visitTypeElement(optionType, data);
    }

    public R visitThisType(@NotNull CjThisType optionType, D data) {
        return visitTypeElement(optionType, data);
    }

    public R visitProperty(@NotNull CjProperty property, D data) {
        return visitNamedDeclaration(property, data);
    }

    public R visitVariable(@NotNull CjVariable variable, D data) {
        return visitNamedDeclaration(variable, data);
    }

    public R visitCallableReferenceExpression(@NotNull CjCallableReferenceExpression expression, D data) {
        return visitDoubleColonExpression(expression, data);
    }

    public R visitDoubleColonExpression(@NotNull CjDoubleColonExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitRangeExpression(@NotNull CjRangeExpression expression, D data) {
        return visitExpression(expression, data);

    }
    public R visitSpawnExpression(@NotNull CjSpawnExpression expression, D data) {
        return visitCallExpression(expression, data);
    }
    public R visitLambdaExpression(@NotNull CjLambdaExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitUnsafeExpression(@NotNull CjUnsafeExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitTupleExpression(@NotNull CjTupleExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitTryExpression(@NotNull CjTryExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitCatchSection(CjCatchClause cjCatchClause, D data) {

        return visitCjElement(cjCatchClause, data);
    }

    public R visitFinallySection(@NotNull CjFinallySection finallySection, D data) {
        return visitCjElement(finallySection, data);
    }

    public R visitMatchExpression(@NotNull CjMatchExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitWhileExpression(@NotNull CjWhileExpression expression, D data) {
        return visitLoopExpression(expression, data);
    }

    public R visitDoWhileExpression(@NotNull CjDoWhileExpression expression, D data) {
        return visitLoopExpression(expression, data);
    }

    public R visitLoopExpression(@NotNull CjLoopExpression loopExpression, D data) {
        return visitExpression(loopExpression, data);
    }

    public R visitForExpression(@NotNull CjForExpression expression, D data) {
        return visitLoopExpression(expression, data);
    }

    public R visitIfExpression(@NotNull CjIfExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitContinueExpression(@NotNull CjContinueExpression expression, D data) {
        return visitExpressionWithLabel(expression, data);
    }

    public R visitBreakExpression(@NotNull CjBreakExpression expression, D data) {
        return visitExpressionWithLabel(expression, data);
    }

    public R visitCollectionLiteralExpression(@NotNull CjCollectionLiteralExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitThrowExpression(@NotNull CjThrowExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitReturnExpression(@NotNull CjReturnExpression expression, D data) {
        return visitExpressionWithLabel(expression, data);
    }

    public R visitParenthesizedExpression(@NotNull CjParenthesizedExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitStringTemplateExpression(@NotNull CjStringTemplateExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitSuperExpression(@NotNull CjSuperExpression expression, D data) {
        return visitExpressionWithLabel(expression, data);
    }

    public R visitThisExpression(@NotNull CjThisExpression expression, D data) {
        return visitExpressionWithLabel(expression, data);
    }

    public R visitExpressionWithLabel(@NotNull CjExpressionWithLabel expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitLiteralStringTemplateEntry(@NotNull CjLiteralStringTemplateEntry entry, D data) {
        return visitStringTemplateEntry(entry, data);
    }

    public R visitEscapeStringTemplateEntry(@NotNull CjEscapeStringTemplateEntry entry, D data) {
        return visitStringTemplateEntry(entry, data);
    }

    public R visitStringTemplateEntry(@NotNull CjStringTemplateEntry entry, D data) {
        return visitCjElement(entry, data);
    }

    public R visitStringTemplateEntryWithExpression(@NotNull CjStringTemplateEntryWithExpression entry, D data) {
        return visitStringTemplateEntry(entry, data);
    }

    public R visitBlockStringTemplateEntry(@NotNull CjBlockStringTemplateEntry entry, D data) {
        return visitStringTemplateEntryWithExpression(entry, data);
    }

    public R visitPostfixExpression(@NotNull CjPostfixExpression expression, D data) {
        return visitUnaryExpression(expression, data);
    }

    public R visitConstantExpression(@NotNull CjConstantExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitValueArgumentList(@NotNull CjValueArgumentList list, D data) {
        return visitCjElement(list, data);
    }

    public R visitArrayAccessExpression(@NotNull CjArrayAccessExpression expression, D data) {
        return visitReferenceExpression(expression, data);
    }

    public R visitSuperTypeEntry(@NotNull CjSuperTypeEntry specifier, D data) {
        return visitSuperTypeListEntry(specifier, data);
    }

    public R visitPrefixExpression(@NotNull CjPrefixExpression expression, D data) {
        return visitUnaryExpression(expression, data);
    }

    public R visitUnaryExpression(@NotNull CjUnaryExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitIsExpression(@NotNull CjIsExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitBinaryExpression(@NotNull CjBinaryExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitSuperTypeCallEntry(@NotNull CjSuperTypeCallEntry call, D data) {
        return visitSuperTypeListEntry(call, data);
    }

    public R visitBinaryWithTypeRHSExpression(@NotNull CjBinaryExpressionWithTypeRHS expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitClassInitializer(@NotNull CjClassInitializer initializer, D data) {
        return visitAnonymousInitializer(initializer, data);
    }

    public R visitAnonymousInitializer(@NotNull CjAnonymousInitializer initializer, D data) {
        return visitDeclaration(initializer, data);
    }

    public R visitSuperTypeListEntry(@NotNull CjSuperTypeListEntry specifier, D data) {
        return visitCjElement(specifier, data);
    }

    public R visitTypeProjection(@NotNull CjTypeProjection typeProjection, D data) {
        return visitCjElement(typeProjection, data);
    }

    public R visitTypeParameterList(@NotNull CjTypeParameterList list, D data) {
        return visitCjElement(list, data);
    }

    public R visitTypeParameter(@NotNull CjTypeParameter parameter, D data) {
        return visitNamedDeclaration(parameter, data);
    }

    public R visitCjFile(@NotNull CjFile file, D data) {
        visitFile(file);
        return null;
    }

    public R visitBlockExpression(@NotNull CjBlockExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitExpression(@NotNull CjExpression expression, D data) {
        return visitCjElement(expression, data);
    }

    public R visitClassBody(@NotNull CjClassBody classBody, D data) {
        return visitCjElement(classBody, data);
    }

    public R visitEnumBody(@NotNull CjEnumBody enumBody, D data) {
        return visitCjElement(enumBody, data);
    }

    public R visitModifierList(@NotNull CjModifierList list, D data) {
        return visitCjElement(list, data);
    }


    public R visitContextReceiverList(@NotNull CjContextReceiverList contextReceiverList, D data) {
        return visitCjElement(contextReceiverList, data);
    }

    public R visitTypeReference(@NotNull CjTypeReference cjTypeReference, D data) {
        return visitCjElement(cjTypeReference, data);
    }


    public R visitSimpleNameExpression(@NotNull CjSimpleNameExpression expression, D data) {
        return visitReferenceExpression(expression, data);
    }

    public R visitReferenceExpression(@NotNull CjReferenceExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitTypeConstraintList(@NotNull CjTypeConstraintList cjTypeConstraintList, D data) {
        return visitCjElement(cjTypeConstraintList, data);
    }

    public R visitTypeConstraint(@NotNull CjTypeConstraint cjTypeConstraint, D data) {
        return visitCjElement(cjTypeConstraint, data);
    }

    public R visitParameterList(@NotNull CjParameterList cjParameterList, D data) {
        return visitCjElement(cjParameterList, data);
    }

    public R visitParameter(CjParameter cjParameter, D data) {
        return visitCjElement(cjParameter, data);
    }

    public R visitDestructuringDeclaration(CjDestructuringDeclaration cjDestructuringDeclaration, D data) {

        return visitCjElement(cjDestructuringDeclaration, data);
    }

    public R visitDestructuringDeclarationEntry(CjDestructuringDeclarationEntry cjDestructuringDeclarationEntry, D data) {
        return visitCjElement(cjDestructuringDeclarationEntry, data);
    }


    public R visitDeclaration(@NotNull CjDeclaration dcl, D data) {
        return visitExpression(dcl, data);
    }

    public R visitTypeAlias(@NotNull CjTypeAlias typeAlias, D data) {
        return visitNamedDeclaration(typeAlias, data);
    }

    public R visitNamedDeclaration(@NotNull CjNamedDeclaration declaration, D data) {
        return visitDeclaration(declaration, data);
    }

    public R visitPrimaryConstructor(@NotNull CjPrimaryConstructor constructor, D data) {
        return visitNamedDeclaration(constructor, data);
    }

    public R visitNamedFunction(CjNamedFunction cjNamedFunction, D data) {
        return visitNamedDeclaration(cjNamedFunction, data);
    }

    public R visitCallExpression(CjCallExpression cjCallExpression, D data) {
        return visitReferenceExpression(cjCallExpression, data);
    }

    public R visitSecondaryConstructor(@NotNull CjSecondaryConstructor constructor, D data) {
        return visitNamedDeclaration(constructor, data);
    }

    public R visitConstructorDelegationCall(CjConstructorDelegationCall cjConstructorDelegationCall, D data) {
        return visitCjElement(cjConstructorDelegationCall, data);
    }

    public R visitPackageDirective(CjPackageDirective cjPackageDirective, D data) {
        return visitCjElement(cjPackageDirective, data);
    }

    public R visitQualifiedExpression(@NotNull CjQualifiedExpression expression, D data) {
        return visitExpression(expression, data);
    }

    public R visitDotQualifiedExpression(@NotNull CjDotQualifiedExpression expression, D data) {
        return visitQualifiedExpression(expression, data);
    }

    public R visitImportDirective(CjImportDirective cjImportDirective, D data) {
        return visitCjElement(cjImportDirective, data);
    }

    public R visitUserType(CjUserType cjUserType, D data) {
        return visitCjElement(cjUserType, data);
    }

    public R visitTypeArgumentList(CjTypeArgumentList cjTypeArgumentList, D data) {
        return visitCjElement(cjTypeArgumentList, data);
    }

    //    public R visitMacroArgumentList(CjMacroAttrArgumentList psi, D data) {
//        return visitCjElement(psi, data);
//    }
    public R visitImportAlias(@NotNull CjImportAlias importAlias, D data) {
        return visitCjElement(importAlias, data);
    }

    public R visitImportList(CjImportList cjImportList, D data) {
        return visitCjElement(cjImportList, data);
    }

    public R visitSuperTypeList(@NotNull CjSuperTypeList list, D data) {
        return visitCjElement(list, data);
    }

    public R visitConstructorCalleeExpression(@NotNull CjConstructorCalleeExpression constructorCalleeExpression, D data) {
        return visitCjElement(constructorCalleeExpression, data);
    }

    public R visitBasicType(CjBasicType cjBasicType, D data) {
        return visitCjElement(cjBasicType, data);
    }

    public R visitKeyword(CjKeyword cjKeyword, D data) {
        return visitCjElement(cjKeyword, data);
    }

    public R visitSimpleNameStringTemplateEntry(@NotNull CjSimpleNameStringTemplateEntry entry, D data) {
        return visitStringTemplateEntryWithExpression(entry, data);
    }

    public R visitEnumEntry(CjEnumEntry cjEnumEntry, D data) {
        return visitTypeStatement(cjEnumEntry, data);

    }

    public R visitMatchEntry(CjMatchEntry cjMatchEntry, D data) {
        return visitCjElement(cjMatchEntry, data);
    }

    public R visitTupleType(@NotNull CjTupleType cjTupleType, D data) {
        return visitCjElement(cjTupleType, data);
    }

    public R visitParenthesizedType(@NotNull CjParenthesizedType cjParenthesizedType, D data) {
        return visitCjElement(cjParenthesizedType, data);
    }

    public R visitFunctionType(@NotNull CjFunctionType type, D data) {
        return visitTypeElement(type, data);
    }

    private R visitTypeElement(@NotNull CjTypeElement type, D data) {
        return visitCjElement(type, data);
    }

    public R visitPropertyAccessor(@NotNull CjPropertyAccessor accessor, D data) {
        return visitDeclaration(accessor, data);
    }

    public R visitExtend(@NotNull CjExtend cjExtend, D data) {
        return visitCjElement(cjExtend, data);
    }

    public R visitClass(@NotNull CjClass cclass, D data) {
        return visitTypeStatement(cclass, data);
    }

    public R visitEnum(@NotNull CjEnum cenum, D data) {
        return visitTypeStatement(cenum, data);
    }

    public R visitStruct(@NotNull CjStruct cstruct, D data) {
        return visitTypeStatement(cstruct, data);
    }

    public R visitInterface(@NotNull CjInterface cinterface, D data) {
        return visitTypeStatement(cinterface, data);
    }


    //    类型声明
    public R visitTypeStatement(@NotNull CjTypeStatement typeStatement, D data) {
        return visitNamedDeclaration(typeStatement, data);
    }

    public R visitAnnotation(@NotNull CjAnnotation annotation, D data) {
        return visitCjElement(annotation, data);
    }

    public R visitMainFunction(CjMainFunction cjMainFunction, D data) {
        return visitDeclaration(cjMainFunction, data);
    }

    public R visitClassInitFunction(@NotNull CjClassInit cjClassInit, D data) {
        return visitDeclaration(cjClassInit, data);
    }
}

File: psi/src/main/kotlin/cn/cangnova/cangjie/psi/CjVisitor.kt
/*
 * Copyright 2024 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */

package cn.cangnova.cangjie.psi

import com.intellij.psi.PsiElementVisitor

/**
 * 访问者模式基类，用于遍历CangJie语言的语法树
 */
open class CjVisitor<R, D> : PsiElementVisitor() {

    open fun visitCjElement(element: CjElement, data: D?): R? {
        visitElement(element)
        return null
    }

    open fun visitQuoteInterpolate(expression: CjQuoteInterpolate, data: D?): R? {
        return visitCjElement(expression, data)
    }

    open fun visitScript(script: CjScript, data: D?): R? {
        return visitDeclaration(script, data)
    }

    /**
     * Required for [CjCommonFile] implementation.
     * It is not expected to be used, because [CjCommonFile] should not be used directly.
     */
    @Suppress("DEPRECATION")
    fun visitCjCommonFile(file: CjCommonFile): R? {
        visitFile(file)
        return null
    }

    open fun visitSynchronizedExpression(expression: CjSynchronizedExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitPatternGuard(expression: CjPatternGuard, data: D?): R? {
        return visitCjElement(expression, data)
    }

    open fun visitSafeQualifiedExpression(expression: CjSafeQualifiedExpression, data: D?): R? {
        return visitQualifiedExpression(expression, data)
    }

    open fun visitMacroExpression(expression: CjMacroExpression, data: D?): R? {
        return visitCjElement(expression, data)
    }

    open fun visitQuoteExpression(element: CjQuoteExpression, data: D?): R? {
        return visitExpression(element, data)
    }

    open fun visitCasePattern(element: CjCasePattern, data: D?): R? {
        return visitCjElement(element, data)
    }

    open fun visitPatternByConstant(element: CjConstantPattern, data: D?): R? {
        return visitCasePattern(element, data)
    }

    open fun visitPatternByBinding(element: CjBindingPattern, data: D?): R? {
        return visitCasePattern(element, data)
    }

    open fun visitMatchConditionWithExpression(condition: CjMatchConditionWithExpression, data: D?): R? {
        return visitCasePattern(condition, data)
    }

    open fun visitPatternByType(element: CjTypePattern, data: D?): R? {
        return visitCasePattern(element, data)
    }

    open fun visitPatternByTuple(element: CjTuplePattern, data: D?): R? {
        return visitCasePattern(element, data)
    }

    open fun visitPatternByWildcard(element: CjWildcardPattern, data: D?): R? {
        return visitCasePattern(element, data)
    }

    open fun visitPatternByEnum(element: CjEnumPattern, data: D?): R? {
        return visitCasePattern(element, data)
    }

    open fun visitOptionType(optionType: CjOptionType, data: D?): R? {
        return visitTypeElement(optionType, data)
    }

    open fun visitVArrayType(optionType: CjVArrayType, data: D?): R? {
        return visitTypeElement(optionType, data)
    }

    open fun visitThisType(optionType: CjThisType, data: D?): R? {
        return visitTypeElement(optionType, data)
    }

    open fun visitProperty(property: CjProperty, data: D?): R? {
        return visitNamedDeclaration(property, data)
    }

    open fun visitVariable(variable: CjVariable, data: D?): R? {
        return visitNamedDeclaration(variable, data)
    }

    open fun visitCallableReferenceExpression(expression: CjCallableReferenceExpression, data: D?): R? {
        return visitDoubleColonExpression(expression, data)
    }

    open fun visitDoubleColonExpression(expression: CjDoubleColonExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitRangeExpression(expression: CjRangeExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitSpawnExpression(expression: CjSpawnExpression, data: D?): R? {
        return visitCallExpression(expression, data)
    }

    open fun visitLambdaExpression(expression: CjLambdaExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitUnsafeExpression(expression: CjUnsafeExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitTupleExpression(expression: CjTupleExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitTryResourceList(resource: CjTryResourceList, data: D?): R? {
        return visitCjElement(resource, data)
    }

    open fun visitTryResource(resource: CjTryResource, data: D?): R? {
        return visitCjElement(resource, data)
    }

    open fun visitTryExpression(expression: CjTryExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitCatchSection(cjCatchClause: CjCatchClause, data: D?): R? {
        return visitCjElement(cjCatchClause, data)
    }

    open fun visitFinallySection(finallySection: CjFinallySection, data: D?): R? {
        return visitCjElement(finallySection, data)
    }

    open fun visitMatchExpression(expression: CjMatchExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitLetExpression(expression: CjLetExpression, data: D?): R? {
        return visitCjElement(expression, data)
    }

    open fun visitWhileExpression(expression: CjWhileExpression, data: D?): R? {
        return visitLoopExpression(expression, data)
    }

    open fun visitDoWhileExpression(expression: CjDoWhileExpression, data: D?): R? {
        return visitLoopExpression(expression, data)
    }

    open fun visitLoopExpression(loopExpression: CjLoopExpression, data: D?): R? {
        return visitExpression(loopExpression, data)
    }

    open fun visitForExpression(expression: CjForExpression, data: D?): R? {
        return visitLoopExpression(expression, data)
    }

    open fun visitIfExpression(expression: CjIfExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitContinueExpression(expression: CjContinueExpression, data: D?): R? {
        return visitExpressionWithLabel(expression, data)
    }

    open fun visitBreakExpression(expression: CjBreakExpression, data: D?): R? {
        return visitExpressionWithLabel(expression, data)
    }

    open fun visitCollectionLiteralExpression(expression: CjCollectionLiteralExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitThrowExpression(expression: CjThrowExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitReturnExpression(expression: CjReturnExpression, data: D?): R? {
        return visitExpressionWithLabel(expression, data)
    }

    open fun visitParenthesizedExpression(expression: CjParenthesizedExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitStringTemplateExpression(expression: CjStringTemplateExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitSuperExpression(expression: CjSuperExpression, data: D?): R? {
        return visitExpressionWithLabel(expression, data)
    }

    open fun visitThisExpression(expression: CjThisExpression, data: D?): R? {
        return visitExpressionWithLabel(expression, data)
    }

    open fun visitExpressionWithLabel(expression: CjExpressionWithLabel, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitLiteralStringTemplateEntry(entry: CjLiteralStringTemplateEntry, data: D?): R? {
        return visitStringTemplateEntry(entry, data)
    }

    open fun visitEscapeStringTemplateEntry(entry: CjEscapeStringTemplateEntry, data: D?): R? {
        return visitStringTemplateEntry(entry, data)
    }

    open fun visitStringTemplateEntry(entry: CjStringTemplateEntry, data: D?): R? {
        return visitCjElement(entry, data)
    }

    open fun visitStringTemplateEntryWithExpression(entry: CjStringTemplateEntryWithExpression, data: D?): R? {
        return visitStringTemplateEntry(entry, data)
    }

    open fun visitBlockStringTemplateEntry(entry: CjBlockStringTemplateEntry, data: D?): R? {
        return visitStringTemplateEntryWithExpression(entry, data)
    }

    open fun visitPostfixExpression(expression: CjPostfixExpression, data: D?): R? {
        return visitUnaryExpression(expression, data)
    }

    open fun visitConstantExpression(expression: CjConstantExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitValueArgumentList(list: CjValueArgumentList, data: D?): R? {
        return visitCjElement(list, data)
    }

    open fun visitArrayAccessExpression(expression: CjArrayAccessExpression, data: D?): R? {
        return visitReferenceExpression(expression, data)
    }

    open fun visitSuperTypeEntry(specifier: CjSuperTypeEntry, data: D?): R? {
        return visitSuperTypeListEntry(specifier, data)
    }

    open fun visitPrefixExpression(expression: CjPrefixExpression, data: D?): R? {
        return visitUnaryExpression(expression, data)
    }

    open fun visitUnaryExpression(expression: CjUnaryExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitIsExpression(expression: CjIsExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitBinaryExpression(expression: CjBinaryExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitSuperTypeCallEntry(call: CjSuperTypeCallEntry, data: D?): R? {
        return visitSuperTypeListEntry(call, data)
    }

    open fun visitBinaryWithTypeRHSExpression(expression: CjBinaryExpressionWithTypeRHS, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitClassInitializer(initializer: CjClassInitializer, data: D?): R? {
        return visitAnonymousInitializer(initializer, data)
    }

    open fun visitAnonymousInitializer(initializer: CjAnonymousInitializer, data: D?): R? {
        return visitDeclaration(initializer, data)
    }

    open fun visitSuperTypeListEntry(specifier: CjSuperTypeListEntry, data: D?): R? {
        return visitCjElement(specifier, data)
    }

    open fun visitTypeProjection(typeProjection: CjTypeProjection, data: D?): R? {
        return visitCjElement(typeProjection, data)
    }

    open fun visitTypeParameterList(list: CjTypeParameterList, data: D?): R? {
        return visitCjElement(list, data)
    }

    open fun visitTypeParameter(parameter: CjTypeParameter, data: D?): R? {
        return visitNamedDeclaration(parameter, data)
    }

    open fun visitCjFile(file: CjFile, data: D?): R? {
        visitFile(file)
        return null
    }

    open fun visitBlockExpression(expression: CjBlockExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitExpression(expression: CjExpression, data: D?): R? {
        return visitCjElement(expression, data)
    }

    open fun visitClassBody(classBody: CjAbstractClassBody, data: D?): R? {
        return visitCjElement(classBody, data)
    }

    open fun visitEnumBody(enumBody: CjEnumBody, data: D?): R? {
        return visitCjElement(enumBody, data)
    }

    open fun visitModifierList(list: CjModifierList, data: D?): R? {
        return visitCjElement(list, data)
    }

    open fun visitContextReceiverList(contextReceiverList: CjContextReceiverList, data: D?): R? {
        return visitCjElement(contextReceiverList, data)
    }

    open fun visitTypeReference(cjTypeReference: CjTypeReference, data: D?): R? {
        return visitCjElement(cjTypeReference, data)
    }

    open fun visitSimpleNameExpression(expression: CjSimpleNameExpression, data: D?): R? {
        return visitReferenceExpression(expression, data)
    }

    open fun visitReferenceExpression(expression: CjReferenceExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitTypeConstraintList(cjTypeConstraintList: CjTypeConstraintList, data: D?): R? {
        return visitCjElement(cjTypeConstraintList, data)
    }

    open fun visitTypeConstraint(cjTypeConstraint: CjTypeConstraint, data: D?): R? {
        return visitCjElement(cjTypeConstraint, data)
    }

    open fun visitParameterList(cjParameterList: CjParameterList, data: D?): R? {
        return visitCjElement(cjParameterList, data)
    }

    open fun visitParameter(cjParameter: CjParameter, data: D?): R? {
        return visitCjElement(cjParameter, data)
    }

    open fun visitDestructuringDeclaration(cjDestructuringDeclaration: CjDestructuringDeclaration, data: D?): R? {
        return visitCjElement(cjDestructuringDeclaration, data)
    }

    open fun visitDestructuringDeclarationEntry(
        cjDestructuringDeclarationEntry: CjDestructuringDeclarationEntry,
        data: D?
    ): R? {
        return visitCjElement(cjDestructuringDeclarationEntry, data)
    }

    open fun visitDeclaration(dcl: CjDeclaration, data: D?): R? {
        return visitExpression(dcl, data)
    }

    open fun visitTypeAlias(typeAlias: CjTypeAlias, data: D?): R? {
        return visitNamedDeclaration(typeAlias, data)
    }

    open fun visitNamedDeclaration(declaration: CjNamedDeclaration, data: D?): R? {
        return visitDeclaration(declaration, data)
    }

    open fun visitPrimaryConstructor(constructor: CjPrimaryConstructor, data: D?): R? {
        return visitNamedDeclaration(constructor, data)
    }

    open fun visitMacroDeclaration(macroDeclaration: CjMacroDeclaration, data: D?): R? {
        return visitNamedDeclaration(macroDeclaration, data)
    }

    open fun visitNamedFunction(cjNamedFunction: CjNamedFunction, data: D?): R? {
        return visitNamedDeclaration(cjNamedFunction, data)
    }

    open fun visitCallExpression(cjCallExpression: CjCallExpression, data: D?): R? {
        return visitReferenceExpression(cjCallExpression, data)
    }

    open fun visitEndSecondaryConstructor(constructor: CjEndSecondaryConstructor, data: D?): R? {
        return visitNamedDeclaration(constructor, data)
    }

    open fun visitSecondaryConstructor(constructor: CjSecondaryConstructor, data: D?): R? {
        return visitNamedDeclaration(constructor, data)
    }

    open fun visitConstructorDelegationCall(cjConstructorDelegationCall: CjConstructorDelegationCall, data: D?): R? {
        return visitCjElement(cjConstructorDelegationCall, data)
    }

    open fun visitPackageDirective(cjPackageDirective: CjPackageDirective, data: D?): R? {
        return visitCjElement(cjPackageDirective, data)
    }

    open fun visitQualifiedExpression(expression: CjQualifiedExpression, data: D?): R? {
        return visitExpression(expression, data)
    }

    open fun visitDotQualifiedExpression(expression: CjDotQualifiedExpression, data: D?): R? {
        return visitQualifiedExpression(expression, data)
    }

    open fun visitImportDirectiveItem(importDirectiveItem: CjImportDirectiveItem, data: D?): R? {
        return visitCjElement(importDirectiveItem, data)
    }

    open fun visitImportDirective(cjImportDirective: CjImportDirective, data: D?): R? {
        return visitCjElement(cjImportDirective, data)
    }

    open fun visitUserType(cjUserType: CjUserType, data: D?): R? {
        return visitCjElement(cjUserType, data)
    }

    open fun visitTypeArgumentList(cjTypeArgumentList: CjTypeArgumentList, data: D?): R? {
        return visitCjElement(cjTypeArgumentList, data)
    }

    open fun visitImportAlias(importAlias: CjImportAlias, data: D?): R? {
        return visitCjElement(importAlias, data)
    }

    open fun visitImportList(cjImportList: CjImportList, data: D?): R? {
        return visitCjElement(cjImportList, data)
    }

    open fun visitSuperTypeList(list: CjSuperTypeList, data: D?): R? {
        return visitCjElement(list, data)
    }

    open fun visitConstructorCalleeExpression(
        constructorCalleeExpression: CjConstructorCalleeExpression,
        data: D?
    ): R? {
        return visitCjElement(constructorCalleeExpression, data)
    }

    open fun visitBasicType(cjBasicType: CjBasicType, data: D?): R? {
        return visitCjElement(cjBasicType, data)
    }

    open fun visitKeyword(cjKeyword: CjKeyword, data: D?): R? {
        return visitCjElement(cjKeyword, data)
    }

    open fun visitSimpleNameStringTemplateEntry(entry: CjSimpleNameStringTemplateEntry, data: D?): R? {
        return visitStringTemplateEntryWithExpression(entry, data)
    }

    open fun visitEnumEntry(cjEnumEntry: CjEnumEntry, data: D?): R? {
        return visitTypeStatement(cjEnumEntry, data)
    }

    open fun visitMatchEntry(cjMatchEntry: CjMatchEntry, data: D?): R? {
        return visitCjElement(cjMatchEntry, data)
    }

    open fun visitTupleType(cjTupleType: CjTupleType, data: D?): R? {
        return visitCjElement(cjTupleType, data)
    }

    open fun visitParenthesizedType(cjParenthesizedType: CjParenthesizedType, data: D?): R? {
        return visitCjElement(cjParenthesizedType, data)
    }

    open fun visitFunctionType(type: CjFunctionType, data: D?): R? {
        return visitTypeElement(type, data)
    }

    protected open fun visitTypeElement(type: CjTypeElement, data: D?): R? {
        return visitCjElement(type, data)
    }

    open fun visitPropertyAccessor(accessor: CjPropertyAccessor, data: D?): R? {
        return visitDeclaration(accessor, data)
    }

    open fun visitExtend(cjExtend: CjExtend, data: D?): R? {
        return visitTypeStatement(cjExtend, data)
    }

    open fun visitClass(klass: CjClass, data: D?): R? {
        return visitTypeStatement(klass, data)
    }

    open fun visitEnum(cenum: CjEnum, data: D?): R? {
        return visitTypeStatement(cenum, data)
    }

    open fun visitStruct(cstruct: CjStruct, data: D?): R? {
        return visitTypeStatement(cstruct, data)
    }

    open fun visitInterface(cinterface: CjInterface, data: D?): R? {
        return visitTypeStatement(cinterface, data)
    }

    // 类型声明
    open fun visitTypeStatement(typeStatement: CjTypeStatement, data: D?): R? {
        return visitNamedDeclaration(typeStatement, data)
    }

    open fun visitAnnotation(annotation: CjAnnotation, data: D?): R? {
        return visitCjElement(annotation, data)
    }

    open fun visitMainFunction(cjMainFunction: CjMainFunction, data: D?): R? {
        return visitDeclaration(cjMainFunction, data)
    }

    open fun visitClassInitFunction(cjClassInit: CjClassInit, data: D?): R? {
        return visitDeclaration(cjClassInit, data)
    }
} 